src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/RandomNumberGenerator.cs

```diff
@@ -165,6 +166,197 @@ public static byte[] GetBytes(int count)
+        /// </summary>
+        /// <param name="destination">The buffer to receive the characters.</param>
+        /// <param name="lowercase">
+        ///   <see langword="true" /> if the hexadecimal characters should be uppercase; <see langword="true" /> if lowercase.
+        ///   The default is <see langword="false" />.
+        /// </param>
+        /// <remarks>
+        ///   The behavior of this is the same as using <seealso cref="GetItems{T}(ReadOnlySpan{T}, Span{T})" /> and
+        ///   specifying hexadecimal characters as the choices. This implementation is optimized specifically for
+        ///   hexadecimal characters.
+        /// </remarks>
+        public static void GetHexString(Span<char> destination, bool lowercase = false)
+        {
+            if (destination.IsEmpty)
+                return;
+
+            GetHexStringCore(destination, lowercase);
+        }
+
+        /// <summary>
+        ///   Creates a string filled with cryptographically random hexadecimal characters.
+        /// </summary>
+        /// <param name="stringLength">The length of string to create.</param>
+        /// <param name="lowercase">
+        ///   <see langword="true" /> if the hexadecimal characters should be uppercase; <see langword="true" /> if lowercase.
+        ///   The default is <see langword="false" />.
+        /// </param>
+        /// <returns>A string populated with random hexadecimal characters.</returns>
+        /// <remarks>
+        ///   The behavior of this is the same as using <seealso cref="GetString" /> and
+        ///   specifying hexadecimal characters as the choices. This implementation is optimized specifically for
+        ///   hexadecimal characters.
+        /// </remarks>
+        public static string GetHexString(int stringLength, bool lowercase = false)
+        {
+            if (stringLength == 0)
+                return string.Empty;
+
+            return string.Create(stringLength, lowercase, GetHexStringCore);
+        }
+
+        /// <summary>
+        ///   Performs an in-place shuffle of a span using cryptographically random number generation.
+        /// </summary>
+        /// <param name="values">The span to shuffle.</param>
+        /// <typeparam name="T">The type of span.</typeparam>
+        public static void Shuffle<T>(Span<T> values)
+        {
+            int n = values.Length;
+
+            for (int i = 0; i < n - 1; i++)
+            {
+                int j = GetInt32(i, n);
+                T temp = values[i];
+                values[i] = values[j];
+                values[j] = temp;
+            }
+        }
+
+        private static void GetHexStringCore(Span<char> destination, bool lowercase)
+        {
+            Debug.Assert(!destination.IsEmpty);
+
+            Span<char> remaining = destination;
+            const int RandomBufferSize = 64; // If this changes, the tests need to be updated since they try to exercise boundary conditions.
```

Comment:
How did you decide on this size?  We expect it'll be common for calls to request >= 128 hex chars?

Reply:
> There's a non-trivial amount of code in support of that optimization... I'm not sure it's worth it.

I chose 128 from personal experience but it's arbitrary. I'm not sure I understand the "nontrivial amount of code in support" though. Regardless of the optimization or not, we want to get some finite amount of bytes in a loop to populate the destination which is an unknown length.

Reply:
> I'm not sure I understand the "nontrivial amount of code in support" though. Regardless of the optimization or not, we want to get some finite amount of bytes in a loop to populate the destination which is an unknown length.

The code as written has three sections:
- Handle 1+ full blocks
- Handle less than a block
- Handle a remaining char

If you're not focused on optimizing for full blocks, you can get rid of one of those sections, e.g.
```C#
private static void GetHexStringCore(Span<char> destination, bool lowercase)
{
    if (!destination.IsEmpty)
    {
        Span<byte> randomBytes = stackalloc byte[64];

        if (destination.Length % 2 != 0)
        {
            int r = RandomNumberGenerator.GetInt32(0, 16);
            destination[0] = (char)(r < 10 ? r + '0' : r + (lowercase ? 'a' : 'A') - 10);
            destination = destination.Slice(1);
        }

        while (!destination.IsEmpty)
        {
            int n = Math.Min(destination.Length / 2, randomBytes.Length);
            RandomNumberGenerator.Fill(randomBytes.Slice(0, n));
            HexConverter.EncodeToUtf16(randomBytes.Slice(0, n), destination, lowercase ? Casing.Lower : Casing.Upper);
            destination = destination.Slice(n * 2);
        }
}
```

Reply:
The thing I was really optimizing for is minimizing the number of times Fill is call (either directly or by GetInt32). The thing about RandomNumberGenerator.Fill is that the number of invocations to it, not the number of bytes extracted, makes a big perf impact on MacOS and Linux when dealing with < 1 KB buffers. Asking it for 1 byte two times is nearly twice as expensive ask asking for 256 bytes just once .

Using a destination length of 5 as an example, you are hitting the RNG twice. The implementation as-is does it once.

We could do-away with caring about whole blocks here, but it's currently been written to balance a number of things, more notably minimizing calls to the RNG. But I I thought it would be worthwhile to share my intention here.

Reply:
> The thing I was really optimizing for is minimizing the number of times Fill is call (either directly or by GetInt32).

Ok.  Still doesn't require the dedicated handling for multiple large blocks, e.g. untested:
```C#
private static void GetHexStringCore(Span<char> destination, bool lowercase)
{
    Span<byte> stack = stackalloc byte[256];
    RandomNumberGenerator.Fill(stack);

    // Handle an odd char, as HexConverter.EncodeToUtf16 requires an even number.
    Span<byte> bytes = stack;
    if (destination.Length % 2 != 0)
    {
        destination[0] = lowercase ? HexConverter.ToCharLower(bytes[0]) : HexConverter.ToCharUpper(bytes[0]);
        destination = destination.Slice(1);
        bytes = bytes.Slice(1);
    }

    while (!destination.IsEmpty)
    {
        if (bytes.IsEmpty)
        {
            RandomNumberGenerator.Fill(stack);
            bytes = stack;
        }

        int n = Math.Min(destination.Length / 2, bytes.Length);
        HexConverter.EncodeToUtf16(bytes.Slice(0, n), destination, lowercase ? Casing.Lower : Casing.Upper);
        destination = destination.Slice(n * 2);
        bytes = bytes.Slice(n);
    }
}
```

I'm simply saying the current implementation appears to be optimizing for really large requests and requiring extra code to do so.  If you believe such large requests are actually common and important to optimize for, that's fine.  I'm simply surprised that's the case.

Reply:
Without having an opinion on the stackalloc size, I think Steve's second version is a bit easier to read than the blocks+odd+rest flow, and correctly involves `ceil(ceil(destination.Length / 2) / blocksize)` calls into RNG.Fill

Reply:
> appears to be optimizing for really large requests
 
Well, that wasn't my intention exactly. My intention was to optimize for small inputs. But if we have a big input, then we need to turn it in to a small input. So we do that 6 lines of a `for` loop to burn down large things to get back on our small ( <= 128 characters) path.

When I [adapt your example](https://gist.github.com/vcsjones/08e5db76b44f968c20ba95e2232812d5) and fix the "overfilling the buffer, it comes out to more lines of code (excluding comments) and similar performance characteristics. I know lines of code doesn't mean much, but, it also needs `Math.Min` in a few places, and I guess just isn't how my brain works.

Here was the benchmark results.

|        Method | size |       Mean |    Error |   StdDev | Allocated |
|-------------- |----- |-----------:|---------:|---------:|----------:|
|       FillHex |    8 |   122.9 ns |  0.13 ns |  0.12 ns |         - |
| FillHexSimple |    8 |   123.0 ns |  0.13 ns |  0.11 ns |         - |
|       FillHex |   11 |   132.5 ns |  0.10 ns |  0.09 ns |         - |
| FillHexSimple |   11 |   133.9 ns |  0.12 ns |  0.10 ns |         - |
|       FillHex |   12 |   130.8 ns |  0.12 ns |  0.11 ns |         - |
| FillHexSimple |   12 |   133.2 ns |  0.66 ns |  0.55 ns |         - |
|       FillHex |   63 |   597.7 ns |  2.38 ns |  2.23 ns |         - |
| FillHexSimple |   63 |   595.5 ns |  0.84 ns |  0.70 ns |         - |
|       FillHex |   64 |   594.3 ns |  1.49 ns |  1.39 ns |         - |
| FillHexSimple |   64 |   608.2 ns | 11.22 ns | 10.50 ns |         - |

I don't mind using one approach or another, but I think they are just two different approaches to solving it.

Reply:
> I think Steve's second version is a bit easier to read than the blocks+odd+rest flow,

Okay. Will adjust.